docs/design/metainfo-rpc.md

```diff
@@ -0,0 +1,129 @@
+
+```protobuf
+message AddressedOrderLimit {
+    orders.OrderLimit2 limit = 1;
+    node.NodeAddress storage_node_address = 2;
+}
+
+service Metainfo {
+    rpc StartStreamAndStartSegment(StartStreamAndStartSegmentRequest) returns (StartStreamAndStartSegmentResponse);
+    rpc CommitSegmentAndStartSegment(CommitSegmentAndStartSegmentRequest) returns (CommitSegmentAndStartSegmentResponse);
+    rpc CommitSegmentAndCommitStream(CommitSegmentAndCommitStreamRequest) returns (CommitSegmentAndCommitStreamResponse);
+
+    // reads / deletes
+}
+
+message StartStreamAndStartSegmentRequest {
+    // Start Stream
+    bytes bucket = 1;
+    bytes path = 2;
+    int64 stream_id = 3; // version, etc. -1 for server to pick.
+
+    // Start Segment
+    int64 segment_index = 4;
+    int64 part_number = 5;
+
+    // the following copied from an earlier version of these requests
+    pointerdb.RedundancyScheme redundancy = 6;
+    int64 max_encrypted_segment_size = 7;
+    google.protobuf.Timestamp expiration = 8;
+}
+
+message CommitSegmentAndStartSegmentRequest {
+    // Commit Segment
+    bytes segment_upload_id = 1;
+    pointerdb.Pointer pointer = 2;
+    repeated orders.OrderLimit2 original_limits = 3;
+
+    // Start Segment
+    bytes stream_upload_id = 4;
+    int64 segment_index = 5;
+    int64 part_number = 6;
+
+    // The following copied from an earlier version of these requests
+    pointerdb.RedundancyScheme redundancy = 7;
+    int64 max_encrypted_segment_size = 8;
+    google.protobuf.Timestamp expiration = 9;
+}
+
+message CommitSegmentAndCommitStreamRequest {
+    // Commit Segment
+    bytes segment_upload_id = 1;
+    pointerdb.Pointer pointer = 2;
+    repeated orders.OrderLimit2 original_limits = 3;
+
+    // Commit Stream
+    bytes stream_upload_id = 4;
+}
+
+message StartStreamAndStartSegmentResponse {
+    // Start Stream
+    bytes stream_upload_id = 1;
+
+    // Start Segment
+    bytes segment_upload_id = 2;
+    repeated AddressedOrderLimit addressed_limits = 3;
+    bytes root_piece_id = 4 [(gogoproto.customtype) = "PieceID", (gogoproto.nullable) = false];
+}
+
+message CommitSegmentAndStartSegmentResponse {
+    // Commit Segment
+    pointerdb.Pointer pointer = 1;
+
+    // Start Segment
+    bytes segment_upload_id = 2;
+    repeated AddressedOrderLimit addressed_limits = 3;
+    bytes root_piece_id = 4 [(gogoproto.customtype) = "PieceID", (gogoproto.nullable) = false];
+}
+
+message CommitSegmentAndCommitStreamResponse {
+    // Commit Segment
+    pointerdb.Pointer pointer = 1;
+
+    // Commit Stream
+}
+```
+
+## Rationale
+
+Alternatives include individual messages (e.g. StartStream, CommitStream, StartSegment, CommitSegment), but this would result in more rpc calls overall.
+Also, trying to combine a StartStream message and StartSegment message would lead to unused fields sometimes, like stream_upload_id.
```

Comment:
Maybe this line would be more clear if it included

```protobuf
message StartStreamAndStartSegment {
	StartStream start_stream = 1;
	StartSegment start_segment = 2;
}
```

as an example of what is meant by combining.

Reply:
I'm not very clear with the added example, the test mention _stream_upload_id_ however the _message_ example doesn't contain it.

I don't bear in mind the current implementation, I've just read this document and trying to understand it with that much wider context.